core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala

```diff
@@ -282,8 +287,41 @@ class ProducerStateManagerTest extends JUnitSuite {
     stateManager.takeSnapshot()
 
     // Check that file exists and it is not empty
-    assertEquals("Directory doesn't contain a single file as expected", 1, logDir.list().length)
-    assertTrue("Snapshot file is empty", logDir.list().head.length > 0)
+    assertEquals(2, logDir.list().length)
+    assertEquals(Some(0), stateManager.oldestSnapshotOffset)
+    assertEquals(Some(2), stateManager.latestSnapshotOffset)
+  }
+
+  @Test
+  def testTakeEmptySnapshot(): Unit = {
+    val epoch = 0.toShort
+    append(stateManager, producerId, epoch, seq = 0, 0L)
+    append(stateManager, producerId, epoch, seq = 1, 1L)
+    stateManager.takeSnapshot()
+
+    stateManager.takeEmptySnapshot(0L)
+    assertEquals(Some(0L), stateManager.oldestSnapshotOffset)
+
+    stateManager.truncateAndReload(0L, 1L, time.milliseconds())
+    assertEquals(0, stateManager.activeProducers.size)
+    assertEquals(0, stateManager.mapEndOffset)
+    assertEquals(Some(0L), stateManager.oldestSnapshotOffset)
+    assertEquals(Some(0L), stateManager.latestSnapshotOffset)
+  }
+
+  @Test
+  def testSnapshotAfterTruncation(): Unit = {
+    val epoch = 0.toShort
+    append(stateManager, producerId, epoch, 0, 0L)
+    append(stateManager, producerId, epoch, 1, 1L)
+    stateManager.takeSnapshot()
+    assertEquals(Some(2), stateManager.latestSnapshotOffset)
+
+    stateManager.truncateAndReload(0L, 1L, time.milliseconds())
+    assertEquals(None, stateManager.latestSnapshotOffset)
```

Comment:
This is related to Ismael's suggestion. We could push the log start offset snapshot invariant into `ProducerStateManager`, but at the moment it is maintained in `Log`. So in fact, the oldest snapshot would also be `None`. For now I will add an assertion while I consider the potential refactor.